### DISCUSSION TITLE
Typeclass-based generalization of computable equality

### DISCUSSION THREAD

**Discussion by hacklex** [FStarLang/FStar#2588] [Ideas]
Since the grand goal, the quotient types, is yet out-of-reach, I thought that perhaps we may introduce something simpler...

[Check out this little example](https://gist.github.com/hacklex/b34b56e4ea534b1ee38f4c2ed3207754).

What could go wrong? 
What do you guys think about this?

**Comment by gancherj** [comment]
The main question I have regarding this is: are there developments in F* out there already which use setoids? I am also interested how the design decisions play out with SMT.

The closest thing I can think of are special notions of equality, like `Seq.equal`. Even though Seq.equal is judgmentally the same as equality, are the `eq_elim` and `eq_intro` lemmas used simply as a convenience, or are they indispensable?

**Reply by hacklex** [reply]
Currently there is FStar.Algebra.CommMonoid.Equiv for example. 

No support from the language side yet I think.

I'd love to have quotient types readily available of course, but those seem to be too much trouble to implement...

**Comment by hacklex** [comment]
Currently I'm experimenting with typeclass-based abstract algebra structures:

* [equatable class](https://github.com/hacklex/CuteCAS/blob/master/FStar.Algebra.Classes.Equatable.fst)
* [grouplike structures](https://github.com/hacklex/CuteCAS/blob/master/FStar.Algebra.Classes.Grouplikes.fst) (semigroups, monoids, ...)
* [ringlike structures](https://github.com/hacklex/CuteCAS/blob/master/FStar.Algebra.Classes.Ringlikes.fst) (rings, domains, fields...)

As you can see in ringlikes, proofs become easier to read when you use (+) and (*) to denote the operations.

They can of course be made way less verbose if I introduce forall opaques like I did in my record-based abstract algebra implementation.

I'm still unfamiliar with tactics, so I can't tell if I can perform trivial transformations with tactics when I use custom equivalence relation -- if anyone gives me some advise on this matter, I'd be grateful.
